perf(api,dispatch-service): move downloading COMBINE archive and uploading to S3 to dispatch service

jonrkarr opened the pull request:
closes #4094

jonrkarr committed:
perf(api,dispatch-service): move downloading COMBINE archive and uploading to S3 to dispatch service

bilalshaikh42 commented on apps/dispatch-service/src/app/submission/dispatch.processor.ts:
```diff
@@ -91,4 +186,39 @@ export class DispatchProcessor {
     this.monitorQueue.add('monitor', monitorData, monitorOptions);
   }
+
+  private getRetryBackoff(): <T>(source: Observable<T>) => Observable<T> {
```
I don't think we should add retrybackoff here. Ideally the uploading of the archive would be done as a separate workflow step. If this step gets into a retry loop for a while, it will cause all the job submissions behind it to also get stuck.

bilalshaikh42 commented on apps/dispatch-service/src/app/submission/dispatch.processor.ts:
```diff
@@ -28,6 +42,87 @@ export class DispatchProcessor {
 
     this.logger.debug(`Starting job for simulation run '${data.runId}' ...`);
 
+    // resolve COMBINE archive and save to S3
```
Throwing these errors here is not a good idea unless they are being caught by the  The API has a filter to catch these and return the proper https response. But the dispatch service is not returning anything, so these would end up as un handled promise rejections that cause the container to exit. This is the same issue that is causing the error with ssh.

bilalshaikh42 commented on apps/dispatch-service/src/app/submission/dispatch.processor.ts:
```diff
@@ -57,8 +152,8 @@ export class DispatchProcessor {
       }
 
       throw new BiosimulationsException(
-        500,
```
This is the error I was referring to that is probably the cause of the ssh issue. I haven't gotten to it yet, but this also needs to be replaced by the queue error handling logic

bilalshaikh42 reviewed (changes requested):
I think this should be architecture some thing like this: 

- Dispatch queue. Checks if url provided. If so, add a job to the upload queue. Else add a job to the HPC queue. Basic validation of incoming message. 

-  Upload queue. Uploads archive to s3, patches simulation with url, then adds job to HPC queue. 

- HPC queue. Submits the sbatch script 

This keeps all the steps atomic and idempotent allowing the error handling to be managed entirely by the queue system.

We'd then attach an error event listener to the HPC and upload queues that would mark the simulation run  as failed if either workflow step fails.

jonrkarr committed:
refactor(api,dispatch-service): separated archive resolution and saving to separate queue

jonrkarr commented:
I moved the archive downloading and saving to S3 to a separate queue. I removed retrying because that will now be handled by Bull.

This is architecture:
- API
  - Uploaded archive: saves to S3 and submits to dispatch queue
  - URL for archive: submits to resolveCombineArchive queue
- resolveCombineArchive queue 
  1. Downloads archive
  2. Saves archive to S3
  3. Submits to dispatch queue
- dispatch queue (same as before)
  1. Submits job to HPC 
  2. Submits to monitoring queue

I'm not sure how you intend to handle marking jobs as failed. I mimicked how the dispatch processor handles this:
- Save failed status to database when maximum number of retries is reached
- Throw exception

bilalshaikh42 committed:
refactor(dispatch-service): add scheduler for combine archive resolver queue

bilalshaikh42 committed:
refactor(api): simplify submission queues into 1 queue, add types

changed the api to just use the submission queue instead of sepreately use dispatch and resolve
combine queues. Adding varying types for submission, that extend base type.

bilalshaikh42 committed:
refactor(dispatch-service): add a processor for submission of jobs

This is mostly to reduce the repitition of job types and queue options across the apps. The API now
just has one queue to deal with

bilalshaikh42 committed:
test(api): fix test for controller

biosimulations-daemon commented:
:tada: This PR is included in version 9.5.1 :tada:

The release is available on [GitHub release](https://github.com/biosimulations/biosimulations/releases/tag/v9.5.1)

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:
